File: tmp/leetcode75-c/65.c
int characterReplacement(char * s, int k){
    int ret = 0;
    int slen = strlen(s);
    int cnt[26] = {0}, max_cnt = 0;
    for (int l = 0, r = 0; r < slen; ++r) {
        cnt[s[r]-'A'] += 1;
        max_cnt = (cnt[s[r]-'A'] > max_cnt) ? cnt[s[r]-'A'] : max_cnt;
        if (r-l+1-max_cnt > k) {
            cnt[s[l]-'A'] -= 1;
            ++l;
        } else 
            ret = (r-l+1 > ret) ? r-l+1 : ret;
    }
    return ret;
}

File: tmp/leetcode75/65.py
'''

CCCAAAA k=3 => 7
    k left over go back

ABBAAAA k=1 => 5

ABBAABBBBB k=2 => 9
ABCAABBBBB

ALgorithm: 2 pointers
'''

class Solution:
    # Sliding Window O(n) time O(n) space
    def characterReplacement(self, s: str, k: int) -> int:
        ret = 0
        cnt = defaultdict(int)
        max_cnt = 0
        l = 0
        for r in range(len(s)):
            cnt[s[r]] += 1
            max_cnt = max(max_cnt, cnt[s[r]])
            
            if r-l+1-max_cnt > k:
                cnt[s[l]] -= 1
                l += 1
            else:
                ret = max(ret, r-l+1)
        return ret

    # O(n^2) Bruteforce
    def characterReplacement(self, s: str, k: int) -> int:
        ret = 0
        count = k
        l, r = 0, 0
        while r < len(s):
            if s[r] != s[l]:
                count -= 1
                if count == -1:
                    count = k
                    ret = max(ret, r-l)
                    l += 1; r = l
            r += 1

        while l > 0 and count > 0:
            count -= 1
            l -= 1    
        ret = max(ret, r-l)
        return ret
